File: backend/app/schemas/comparison.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class MissingCardRead(BaseModel):
    name: str
    oracle_id: str | None
    required_quantity: int
    owned_quantity: int
    missing_quantity: int


class ComparisonRowErrorRead(BaseModel):
    row_number: int
    raw: dict[str, Any]
    error: str


class ComparisonResponse(BaseModel):
    mode: str
    total_required_cards: int
    total_required_quantity: int
    total_owned_quantity: int
    coverage_percent: float
    is_fully_buildable: bool
    missing: list[MissingCardRead]
    row_errors: list[ComparisonRowErrorRead]

File: frontend/src/types/comparison.ts
export type ComparisonMode = "oracle" | "printing";
export type DecklistSourceType = "text_list" | "json" | "generic_csv";

export interface MissingCard {
  name: string;
  oracle_id: string | null;
  required_quantity: number;
  owned_quantity: number;
  missing_quantity: number;
}

export interface ComparisonRowError {
  row_number: number;
  raw: Record<string, unknown>;
  error: string;
}

export interface ComparisonResponse {
  mode: ComparisonMode;
  total_required_cards: number;
  total_required_quantity: number;
  total_owned_quantity: number;
  coverage_percent: number;
  is_fully_buildable: boolean;
  missing: MissingCard[];
  row_errors: ComparisonRowError[];
}
